### PR TITLE
Add flag to output error if a typed function is used with an untyped decorator

### PR DESCRIPTION
**Pull request by ilinum**

### CODE REVIEW

File: mypy/checker.py [hunk 1]
```diff
@@ -2275,6 +2276,26 @@ def visit_decorator(self, e: Decorator) -> None:
...
             self.check_incompatible_property_override(e)
 
+    def check_for_untyped_decorator(self,
+                                    func: FuncDef,
+                                    dec_type: Type,
+                                    dec_expr: Expression) -> None:
+        def is_typed_callable(c: Optional[Type]) -> bool:
+            if not c or not isinstance(c, CallableType):
+                return False
+            return all(not isinstance(t, AnyType) or not t.implicit
```
Comment (ddfisher) [inline mypy/checker.py]: Nit: I think this would be much clearer as `not any(isinstance(t, AnyType) and t.implicit)`.

File: mypy/checker.py [hunk 2]
```diff
@@ -2275,6 +2276,26 @@ def visit_decorator(self, e: Decorator) -> None:
         if e.func.is_property:
             self.check_incompatible_property_override(e)
 
+    def check_for_untyped_decorator(self,
+                                    func: FuncDef,
+                                    dec_type: Type,
+                                    dec_expr: Expression) -> None:
+        def is_typed_callable(c: Optional[Type]) -> bool:
```
Comment (ddfisher) [inline mypy/checker.py]: It's better to not define subfunctions unless they actually use variables from the parent function.  I'd make these top-level instead.

File: mypy/checker.py (hunk 1 above)
Comment (ddfisher) [inline mypy/checker.py]: (If you disagree, feel free to leave as-is.)

Comment (ilinum) [inline mypy/checker.py]: Hmm, I think that's better!

File: mypy/checker.py (hunk 2 above)
Comment (ilinum) [inline mypy/checker.py]: All right, that makes sense.

**Comment by gvanrossum** [comment]
The main concern is that this will be seen as a regression by people who have carefully annotated (part of) their codebase but are using untyped decorators (maybe from another part of their codebase). On the one hand these people have a problem (unannotated code persists, and the call sites may have further errors), on the other hand they may not have the bandwidth to solve it quickly, and until it's solved they can no longer use the `--disallow-untyped-defs` flag at all to require that all functions to be annotated. Which is why I would prefer this to be a separate flag (e.g. `--disallow-untyped-decorators`).

**Comment by ilinum** [comment]
After thinking about this some more, I agree with you, this change would discourage use of `--disallow-untyped-defs`, which is something we definitely we want to avoid.

I think it would be reasonable to have a new flag `--disallow-untyped-decorators`. 

Just a note that there is `--disallow-any=decorator` (https://github.com/python/mypy/commit/460a13beb94b3009b502f80939e7d489caee3e1e) that disallows functions that have ``Any`` in their signature after decorator transformation.

**Comment by gvanrossum** [comment]
OK, let's do the separate flag. Re: `--disallow-any=decorator` -- maybe
that should be `decorated` instead? Or are you suggesting that that flag is
sufficient for this use case? Maybe it is.

**Comment by ddfisher** [comment]
(The flag is actually `--disallow-any=decorated`, which is definitely better than `--disallow-any=decorator`.)

**Comment by ddfisher** [comment]
I think a new flag is the right way forward here.

**Comment by ddfisher** [comment]
(Minor note: PR/commit names should be actions.)

File: mypy/checker.py [hunk 3]
```diff
@@ -3223,3 +3233,16 @@ def push_class(self, info: TypeInfo) -> Iterator[None]:
 @contextmanager
 def nothing() -> Iterator[None]:
     yield
+
+
+def is_typed_callable(c: Optional[Type]) -> bool:
+    if not c or not isinstance(c, CallableType):
+        return False
+    return not any(isinstance(t, AnyType) and t.implicit
```
Comment (ddfisher) [inline mypy/checker.py]: It looks like this is checking for a fully typed callable, but we should check for a callable with that has _any_ annotations.  This case should have appeared in tests.

File: test-data/unit/check-flags.test [hunk 4]
```diff
@@ -44,6 +44,46 @@ def f():
...
+[case testDisallowUntypedDefsUnresolvedDecorator]
+# flags: --disallow-untyped-decorators --ignore-missing-imports
+from nonexistent import d
+
+@d  # E: Untyped decorator makes function "f" untyped
+def f(i: int) -> int:
+    return i
+
+[case testDisallowUntypedDefsMultipleUntypedDecorators]
+# flags: --disallow-untyped-decorators --disallow-untyped-defs
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: What is the `--disallow-untyped-defs` part of this test testing?

File: test-data/unit/cmdline.test [hunk 5]
```diff
@@ -290,6 +290,17 @@ def f(a): pass
 z.py:1: error: Function is missing a type annotation
 x.py:1: error: Function is missing a type annotation
 
+[case testDisallowUntypedDefsUntypedFunc]
```
Comment (ddfisher) [inline test-data/unit/cmdline.test]: Why can't this test be in `check-flags.test`?

File: test-data/unit/check-flags.test [hunk 6]
```diff
@@ -44,6 +44,46 @@ def f():
...
+
+[case testDisallowUntypedDefsUntypedDecoratorUntypedFunc]
+# flags: --disallow-untyped-decorators
+
+def d(p):
+    return p
+
+@d  # no error
+def f(): pass
+
```
Comment (ddfisher) [inline test-data/unit/check-flags.test]: Please add tests for partially typed functions.

File: test-data/unit/check-flags.test (hunk 4 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Hmm, my original thinking was to make sure to interacts well with `--disallow-untyped-defs` but I think that the flags don't have that much in common. I removed the test and added another one for multiple decorators.

File: test-data/unit/check-flags.test (hunk 6 above)
Comment (ilinum) [inline test-data/unit/check-flags.test]: Done!

File: test-data/unit/cmdline.test (hunk 5 above)
Comment (ilinum) [inline test-data/unit/cmdline.test]: Not sure of the original reason.
Moved to `check-flags.test`.

File: mypy/checker.py (hunk 3 above)
Comment (ilinum) [inline mypy/checker.py]: Done!

**Comment by ilinum** [comment]
Okay, code review feedback should be addressed now!

File: mypy/checker.py [hunk 7]
```diff
@@ -3255,8 +3255,8 @@ def nothing() -> Iterator[None]:
 def is_typed_callable(c: Optional[Type]) -> bool:
     if not c or not isinstance(c, CallableType):
         return False
-    return not any(isinstance(t, AnyType) and t.implicit
-                   for t in c.arg_types + [c.ret_type])
+    return any(not isinstance(t, AnyType) or not t.implicit
```
Comment (ddfisher) [inline mypy/checker.py]: Nit: I think this may be slightly clearer as `not all(... and ...)`

Comment (ilinum) [inline mypy/checker.py]: Hmm, yep, I think that's better!
I will make the change, make sure it passes the tests and then merge.

**Pull request merged**